From 00eb00f5aee05dba8e466a3f7346d4582b563408 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Thu, 3 Feb 2011 18:06:11 -0700
Subject: [PATCH 6/7] Add menus for each search type screen and owner lookup ability

Add a new "Search"  top level menu with 4 sub menus, one
for each Search type.  These menus link to a default screen
for each type with an owner lookup box when appropriate.

Change-Id: I6a8d7d83e0a24a8b1ecf98d3e41c9ce1fc316f9e
---
 .../java/com/google/gerrit/common/PageLinks.java   |    5 +
 .../main/java/com/google/gerrit/client/Gerrit.java |    8 +
 .../com/google/gerrit/client/GerritConstants.java  |    6 +
 .../gerrit/client/GerritConstants.properties       |    6 +
 .../google/gerrit/client/searches/OwnerScreen.java |  156 ++++++++++++++++++++
 .../gerrit/client/searches/SearchConstants.java    |    6 +
 .../client/searches/SearchConstants.properties     |    6 +
 .../gerrit/client/searches/SearchScreen.java       |   21 ++-
 8 files changed, 207 insertions(+), 7 deletions(-)
 create mode 100644 gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/OwnerScreen.java

diff --git a/gerrit-common/src/main/java/com/google/gerrit/common/PageLinks.java b/gerrit-common/src/main/java/com/google/gerrit/common/PageLinks.java
index 0f29214..1777c29 100644
--- a/gerrit-common/src/main/java/com/google/gerrit/common/PageLinks.java
+++ b/gerrit-common/src/main/java/com/google/gerrit/common/PageLinks.java
@@ -18,6 +18,7 @@ import com.google.gerrit.common.data.AccountInfo;
 import com.google.gerrit.common.data.ChangeInfo;
 import com.google.gerrit.reviewdb.Account;
 import com.google.gerrit.reviewdb.Change;
+import com.google.gerrit.reviewdb.Owner;
 import com.google.gerrit.reviewdb.Project;
 import com.google.gerrit.reviewdb.Change.Status;
 import com.google.gwtorm.client.KeyUtil;
@@ -61,6 +62,10 @@ public class PageLinks {
     return "q," + KeyUtil.encode(query) + "," + TOP;
   }
 
+  public static String toSearches(final Owner.Type type) {
+    return "searches," + type.getId();
+  }
+
   public static String projectQuery(Project.NameKey proj, Status status) {
     switch (status) {
       case ABANDONED:
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java
index d5ccdf9..55d8514 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java
@@ -30,6 +30,7 @@ import com.google.gerrit.common.data.SystemInfoService;
 import com.google.gerrit.reviewdb.Account;
 import com.google.gerrit.reviewdb.AccountDiffPreference;
 import com.google.gerrit.reviewdb.AccountGeneralPreferences;
+import com.google.gerrit.reviewdb.Owner;
 import com.google.gwt.core.client.EntryPoint;
 import com.google.gwt.core.client.GWT;
 import com.google.gwt.core.client.JavaScriptObject;
@@ -457,6 +458,13 @@ public class Gerrit implements EntryPoint {
       menuLeft.selectTab(0);
     }
 
+    m = new LinkMenuBar();
+    addLink(m, C.menuSearchesUsers(), PageLinks.toSearches(Owner.Type.USER));
+    addLink(m, C.menuSearchesGroups(), PageLinks.toSearches(Owner.Type.GROUP));
+    addLink(m, C.menuSearchesProjects(), PageLinks.toSearches(Owner.Type.PROJECT));
+    addLink(m, C.menuSearchesSite(), PageLinks.toSearches(Owner.Type.SITE));
+    menuLeft.add(m, C.menuSearches());
+
     if (signedIn) {
       m = new LinkMenuBar();
       addLink(m, C.menuGroups(), PageLinks.ADMIN_GROUPS);
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.java
index b32a32d..04c5816 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.java
@@ -59,6 +59,12 @@ public interface GerritConstants extends Constants {
   String menuMyWatchedChanges();
   String menuMyStarredChanges();
 
+  String menuSearches();
+  String menuSearchesUsers();
+  String menuSearchesGroups();
+  String menuSearchesProjects();
+  String menuSearchesSite();
+
   String menuAdmin();
   String menuPeople();
   String menuGroups();
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.properties b/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.properties
index 3363015..dde96ff 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.properties
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/GerritConstants.properties
@@ -42,6 +42,12 @@ menuMyDrafts = Drafts
 menuMyStarredChanges = Starred Changes
 menuMyWatchedChanges = Watched Changes
 
+menuSearches = Searches
+menuSearchesUsers = Users
+menuSearchesGroups = Groups
+menuSearchesProjects = Projects
+menuSearchesSite = Site
+
 menuAdmin = Admin
 menuPeople = People
 menuGroups = Groups
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/OwnerScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/OwnerScreen.java
new file mode 100644
index 0000000..337e942
--- /dev/null
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/OwnerScreen.java
@@ -0,0 +1,156 @@
+// Copyright (C) 2011 The Android Open Source Project
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Code Aurora Forum, Inc. nor the names of its
+//       contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED "AS IS" AND ANY EXPRESS OR IMPLIED
+// WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT
+// ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS
+// BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+// BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
+// OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
+// IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package com.google.gerrit.client.searches;
+
+import com.google.gerrit.client.ui.AccountSuggestOracle;
+import com.google.gerrit.client.ui.AccountGroupSuggestOracle;
+import com.google.gerrit.client.ui.HintTextBox;
+import com.google.gerrit.client.ui.ProjectNameSuggestOracle;
+import com.google.gerrit.client.ui.RPCSuggestOracle;
+import com.google.gerrit.client.ui.Screen;
+import com.google.gerrit.client.ui.SuggestUtil;
+import com.google.gerrit.client.rpc.GerritCallback;
+import com.google.gerrit.common.data.AccountInfo;
+import com.google.gerrit.reviewdb.Owner;
+
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.event.dom.client.KeyCodes;
+import com.google.gwt.event.dom.client.KeyPressEvent;
+import com.google.gwt.event.dom.client.KeyPressHandler;
+import com.google.gwt.event.logical.shared.SelectionEvent;
+import com.google.gwt.event.logical.shared.SelectionHandler;
+import com.google.gwt.user.client.ui.Button;
+import com.google.gwt.user.client.ui.Label;
+import com.google.gwt.user.client.ui.SuggestBox;
+import com.google.gwt.user.client.ui.SuggestOracle;
+import com.google.gwt.user.client.ui.SuggestOracle.Suggestion;
+
+public class OwnerScreen extends Screen {
+  protected Owner.Id ownerId;
+
+  protected HintTextBox owner;
+  protected SuggestBox ownerSug;
+  protected Button browse;
+
+  private boolean submitOnSelection;
+
+  protected OwnerScreen(final Owner.Id id) {
+    super();
+    ownerId = id;
+  }
+
+  @Override
+  protected void onInitUI() {
+    super.onInitUI();
+
+    owner = createOwner();
+    if (owner != null) {
+      ownerSug = new SuggestBox(new RPCSuggestOracle(createOwnerOracle()), owner);
+      ownerSug.addSelectionHandler(new SelectionHandler<Suggestion>() {
+        @Override
+        public void onSelection(SelectionEvent<Suggestion> event) {
+          if (submitOnSelection) {
+            submitOnSelection = false;
+            onBrowse();
+          }
+        }
+      });
+
+      browse = new Button(Util.C.buttonBrowseSearches());
+      browse.addClickHandler(new ClickHandler() {
+        @Override
+          public void onClick(ClickEvent event) {
+            onBrowse();
+          }
+        });
+
+      owner.addKeyPressHandler(new KeyPressHandler() {
+          @Override
+          public void onKeyPress(KeyPressEvent event) {
+            submitOnSelection = false;
+
+            if (event.getCharCode() == KeyCodes.KEY_ENTER) {
+              if (ownerSug.isSuggestionListShowing()) {
+                submitOnSelection = true;
+              } else {
+                onBrowse();
+              }
+            }
+          }
+        });
+
+      add(new Label(Util.C.headingBrowse()));
+      add(ownerSug);
+      add(browse);
+    }
+  }
+
+  protected HintTextBox createOwner() {
+    HintTextBox owner = new HintTextBox();
+    switch(ownerId.getType()) {
+      case USER:    owner.setHintText(Util.C.defaultAccountName());
+        break;
+      case GROUP:   owner.setHintText(Util.C.defaultAccountGroupName());
+        break;
+      case PROJECT: owner.setHintText(Util.C.defaultProjectName());
+        break;
+      case SITE:    owner = null;
+        break;
+    }
+    return owner;
+  }
+
+  protected SuggestOracle createOwnerOracle() {
+    switch(ownerId.getType()) {
+      case USER:    return new AccountSuggestOracle();
+      case GROUP:   return new AccountGroupSuggestOracle();
+      case PROJECT: return new ProjectNameSuggestOracle();
+    }
+    return null;
+  }
+
+  public void onBrowse() {
+    if (ownerId.getType() == Owner.Type.USER) {
+      final String nameEmail = ownerSug.getText();
+      if (nameEmail.length() == 0) {
+        return;
+      }
+      SuggestUtil.SVC.getAccountInfo(nameEmail, new GerritCallback<AccountInfo>() {
+          public void onSuccess(final AccountInfo a) {
+            browse(a.getUserName());
+          }
+        });
+    } else {
+      browse(owner.getText());
+    }
+  }
+
+  public void browse(String ownerForLink) {
+  }
+}
\ No newline at end of file
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.java
index 160facb..d26509d 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.java
@@ -30,14 +30,20 @@ package com.google.gerrit.client.searches;
 import com.google.gwt.i18n.client.Constants;
 
 public interface SearchConstants extends Constants {
+  String defaultAccountName();
+  String defaultAccountGroupName();
+  String defaultProjectName();
   String defaultSearchName();
 
+  String buttonBrowseSearches();
   String buttonDeleteSearches();
   String buttonCancelSearch();
   String buttonInsertNewSearch();
   String buttonInsertSearch();
   String buttonSaveSearch();
 
+  String headingBrowse();
+
   String columnSearchName();
   String columnSearchDescription();
   String columnSearchQuery();
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.properties b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.properties
index 430d9bd..38b88fd 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.properties
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchConstants.properties
@@ -1,11 +1,17 @@
+defaultAccountName = (User)
+defaultAccountGroupName = (Group)
+defaultProjectName = (Project)
 defaultSearchName = (Untitled)
 
+buttonBrowseSearches = Browse
 buttonDeleteSearches = Delete
 buttonCancelSearch = Cancel Edit
 buttonInsertNewSearch = Add New Search
 buttonInsertSearch = Add Search
 buttonSaveSearch = Save Search
 
+headingBrowse = Browse searches for:
+
 columnSearchName = Name
 columnSearchDescription = Description
 columnSearchQuery = Query String
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchScreen.java
index 50bbd90..3367fc9 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/searches/SearchScreen.java
@@ -29,7 +29,6 @@ package com.google.gerrit.client.searches;
 
 import com.google.gerrit.client.Dispatcher;
 import com.google.gerrit.client.Gerrit;
-import com.google.gerrit.client.ui.Screen;
 import com.google.gerrit.client.rpc.GerritCallback;
 import com.google.gerrit.client.rpc.ScreenLoadCallback;
 import com.google.gerrit.common.data.OwnerInfo;
@@ -43,7 +42,7 @@ import com.google.gwt.user.client.ui.Button;
 
 import java.util.List;
 
-public class SearchScreen extends Screen {
+public class SearchScreen extends OwnerScreen {
   protected SearchList list;
   protected SearchTable table;
   protected Search.Key key;
@@ -82,7 +81,7 @@ public class SearchScreen extends Screen {
     };
 
   public SearchScreen(final Search.Key key) {
-    super();
+    super(key.getOwnerId());
     this.key = key;
   }
 
@@ -159,14 +158,18 @@ public class SearchScreen extends Screen {
 
   public void refresh() {
     loadSearchList();
+    if (owner != null) {
+      owner.setFocus(true);
+      owner.setFocus(false); // why do we have to do this to make Hint blurr?
+    }
   }
 
   public void loadSearchList() {
-    Util.SEARCH_LIST_SVC.getSearchList(key.getOwnerId(), cbList);
+    Util.SEARCH_LIST_SVC.getSearchList(ownerId, cbList);
   }
 
   public void loadSearches() {
-    Util.SEARCH_LIST_SVC.getSearches(key.getOwnerId(), cbSearches);
+    Util.SEARCH_LIST_SVC.getSearches(ownerId, cbSearches);
   }
 
   private void displaySearches(List<Search> searches) {
@@ -174,6 +177,10 @@ public class SearchScreen extends Screen {
     table.refresh(key);
   }
 
+  public void browse(String ownerForLink) {
+    History.newItem(Dispatcher.toSearches(ownerId.getType(), ownerForLink));
+  }
+
   protected void display(final SearchList list) {
     this.list = list;
     OwnerInfo owner = list.getOwnerInfo();
@@ -216,14 +223,14 @@ public class SearchScreen extends Screen {
   protected String getSearchTitle() {
     String owner = list.getOwnerInfo().getOwnerForDisplay();
     if ("".equals(owner)) {
-      switch(key.getType()) {
+      switch(ownerId.getType()) {
         case USER:    return Util.C.titleSearchUser();
         case GROUP:   return Util.C.titleSearchGroup();
         case PROJECT: return Util.C.titleSearchProject();
         case SITE:    return Util.C.titleSearchSite();
       }
     } else {
-      switch(key.getType()) {
+      switch(ownerId.getType()) {
         case USER:    return Util.M.titleSearchUser(owner);
         case GROUP:   return Util.M.titleSearchGroup(owner);
         case PROJECT: return Util.M.titleSearchProject(owner);
-- 
1.7.4.1

